app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go

```diff
@@ -46,32 +49,115 @@ const (
 	lbName string
 	hcs    healthcheck.HealthCheckSyncerInterface
+	bss    backendservice.BackendServiceSyncerInterface
 	client kubeclient.Interface
+	igp    ingressig.InstanceGroups
 }
 
 func NewLoadBalancerSyncer(lbName string, client kubeclient.Interface, cloud *gce.GCECloud) *LoadBalancerSyncer {
 	namer := utilsnamer.NewNamer(mciPrefix, lbName)
 	return &LoadBalancerSyncer{
 		lbName: lbName,
 		hcs:    healthcheck.NewHealthCheckSyncer(namer, cloud),
+		bss:    backendservice.NewBackendServiceSyncer(namer, cloud),
 		client: client,
+		igp:    cloud,
 	}
 }
 
 func (l *LoadBalancerSyncer) CreateLoadBalancer(ing *v1beta1.Ingress) error {
+	var err error
 	ports := l.ingToNodePorts(ing)
-	// Create health check to be used by the backend service.
-	if err := l.hcs.EnsureHealthCheck(l.lbName, ports); err != nil {
+	// Create health checks to be used by the backend service.
+	healthChecks, hcErr := l.hcs.EnsureHealthCheck(l.lbName, ports)
+	if hcErr != nil {
+		// Keep aggregating errors and return all at the end, rather than giving up on the first error.
+		err = multierror.Append(err, hcErr)
+	}
+	igs, namedPorts, err := l.getIGsAndNamedPorts(ing)
+	// Cant really create any backend service without named ports. No point continuing.
+	if err != nil {
+		return err
+	}
+	// Create backend service. This should always be called after the health check since backend service needs to point to the health check.
+	if err := l.bss.EnsureBackendService(l.lbName, ports, healthChecks, namedPorts, igs); err != nil {
 		return err
 	}
 	return nil
 }
 
+// Returns links to all instance groups and named ports on any one of them.
+// Note that it picks an instance group at random and returns the named ports for that instance group, assuming that the named ports are same on all instance groups.
+// Also note that this returns all named ports on the instance group and not just the ones relevant to the given ingress.
+func (l *LoadBalancerSyncer) getIGsAndNamedPorts(ing *v1beta1.Ingress) ([]string, backendservice.NamedPortsMap, error) {
+	// Keep trying until ingress gets the instance group annotation.
+	for try := true; try; time.Sleep(1 * time.Second) {
+		// Fetch the ingress from a cluster.
+		ing, err := l.getIng(ing.Name, ing.Namespace)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error %s in fetching ingress", err)
+		}
+		if ing.Annotations == nil || ing.Annotations[instanceGroupsAnnotationKey] == "" {
+			// keep trying
+			fmt.Println("Waiting for ingress to get", instanceGroupsAnnotationKey, "annotation.....")
+			continue
+		}
+		annotationValue := ing.Annotations[instanceGroupsAnnotationKey]
+		glog.V(3).Infof("Found instance groups annotation value: %s", annotationValue)
+		// Get the instance group name.
+		type InstanceGroupsAnnotationValue struct {
+			Name string
+			Zone string
+		}
+		var values []InstanceGroupsAnnotationValue
+		if err := json.Unmarshal([]byte(annotationValue), &values); err != nil {
+			return nil, nil, err
```

Comment:
If you get an error from Unmarshal, it's going to be very low-level. can you add more context with Errorf?

Reply:
Done